### PR TITLE
Pruned transducer stateless2 for AISHELL-1

### PR DESCRIPTION
**Pull request by TeaPoly**
This PR is to implement pruned transducer stateless2 for AISHELL-1.
Results will be synced in the future.

### CODE REVIEW

**Comment by TeaPoly** [comment]
Since there is only AISHELL-1 data, it will run 60~90 epochs to check final results.

Here is current results：

|   | test | dev | comment
-- | -- | -- | --
greedy search | 5.83 | 5.32 | --epoch 26 --avg 5 --max-duration 50

```
export CUDA_VISIBLE_DEVICES="0,1"

./pruned_transducer_stateless2/train.py \
        --world-size 2 \
        --num-epochs 90 \
        --start-epoch 0 \
        --exp-dir pruned_transducer_stateless2/exp \
        --max-duration 200 \

./pruned_transducer_stateless2/decode.py \
        --epoch 26 \
        --avg 5 \
        --exp-dir ./pruned_transducer_stateless2/exp \
        --lang-dir data/lang_char \
        --max-duration 50 \
        --decoding-method greedy_search
```
The tensorboard log is available at https://tensorboard.dev/experiment/NlSqLGbHQjuw0i9q248h6w/

**Comment by csukuangfj** [comment]
Is `-epoch 26 --avg 5 ` the best one? Have you tried other combinations of `--epoch` and `--avg`?

**Comment by TeaPoly** [comment]
> Is `-epoch 26 --avg 5 ` the best one? Have you tried other combinations of `--epoch` and `--avg`?

Not sure. I haven't do a lot of test experiments yet because only 26 epochs run so far.

**Comment by TeaPoly** [comment]
> I suggest that you try different combinations of `--epoch` and `--avg` with `--decoding-method=greedy_search` to find the best one and then switch to `modified_beam_search` and `fast_beam_search` to decode your best combination.
> 
> greedy search is super fast and it won't take you much time to try one combination of `--epoch` and `--avg`.

Thanks for your suggestion, I will do this later.

**Comment by luomingshuang** [comment]
You can combine different `epoch` and `avg` for decoding by building a `run_decode.sh`, which contains the following codes ( you can change the number of `num_epochs` ):
```
export CUDA_VISIBLE_DEVICES="0"
num_epochs=26
for ((i=$num_epochs; i>=0; i--));
do
    for ((j=1; j<=$i; j++));
    do
        python3 pruned_transducer_stateless2/decode.py \
            --exp-dir ./pruned_transducer_stateless2/exp \
            --lang-dir data/lang_char \
            --epoch $i \
            --avg $j \
            --max-duration 100 \
            --decoding-method greedy_search
    done
done
```

**Comment by TeaPoly** [comment]
Currently running to the 47th epoch. It seems that `valid_simple_loss`  is unusual after 30 epochs. 

The tensorboard log is available at https://tensorboard.dev/experiment/AqID3VrzQsmkTya9sG5m7Q/#scalars

Do I need to continue or change some experiment parameters to restart? @csukuangfj @luomingshuang 

Here is current best `greedy_search` for `test` results：

| |test | dev | comment
-- | -- | -- | --
greedy search | 5.53 | 5.13 | --epoch 46 --avg 12 --max-duration 100
modified beam search | 5.35 | 4.9 | --epoch 46 --avg 12 --max-duration 100
fast beam search | 5.51 | 5.09 | --epoch 46 --avg 12 --max-duration 100

```
export CUDA_VISIBLE_DEVICES="0,1"

./pruned_transducer_stateless2/train.py \
        --world-size 2 \
        --num-epochs 90 \
        --start-epoch 0 \
        --exp-dir pruned_transducer_stateless2/exp \
        --max-duration 200 \

for epoch in 46; do
  for avg in 12; do
      for m in greedy_search modified_beam_search fast_beam_search ; do
      ./pruned_transducer_stateless2/decode.py \
        --epoch $epoch \
        --avg $avg \
        --exp-dir ./pruned_transducer_stateless2/exp \
        --lang-dir data/lang_char \
        --max-duration 100 \
        --decoding-method $m

    done
  done
done

```

**Comment by csukuangfj** [comment]
> Do I need to continue or change some experiment parameters to restart?

I think there would be little improvement if you continue to train it for more epochs.

**Comment by luomingshuang** [comment]
Maybe you can have a look at your training tensorboard logs, the validation loss.

**Comment by luomingshuang** [comment]
Yes, according to the results https://github.com/k2-fsa/icefall/blob/master/egs/aishell/ASR/RESULTS.md, you can go on your training process with more epochs. And it might help you get a little improvement.

**Comment by TeaPoly** [comment]
Here is final best result:

|                      | test | dev  | comment                                |
| -------------------- | ---- | ---- | -------------------------------------- |
| greedy search        | 5.20 | 4.78 | --epoch 72 --avg 14 --max-duration 200 |
| modified beam search | 5.07 | 4.63 | --epoch 72 --avg 14 --max-duration 200 |
| fast beam search     | 5.13 | 4.70 | --epoch 72 --avg 14 --max-duration 200 |

Training command is:

```bash
./prepare.sh

export CUDA_VISIBLE_DEVICES="0,1"

./pruned_transducer_stateless2/train.py \
        --world-size 2 \
        --num-epochs 90 \
        --start-epoch 0 \
        --exp-dir pruned_transducer_stateless2/exp \
        --max-duration 200 \
```

The tensorboard log is available at
https://tensorboard.dev/experiment/QI3PVzrGRrebxpbWUPwmkA/

The decoding command is:

```bash
for epoch in 72; do
  for avg in 14; do
      for m in greedy_search modified_beam_search fast_beam_search ; do
      ./pruned_transducer_stateless2/decode.py \
        --epoch $epoch \
        --avg $avg \
        --exp-dir ./pruned_transducer_stateless2/exp \
        --lang-dir data/lang_char \
        --max-duration 200 \
        --decoding-method $m

    done
  done
done
```

Pretrained models, training logs, decoding logs, and decoding results
are available at
<https://huggingface.co/teapoly/icefall-aishell-pruned-transducer-stateless2-2022-08-18>

**Comment by luomingshuang** [comment]
Cool, it seems that the results get a noticeable improvement.

File: egs/aishell/ASR/RESULTS.md [hunk 1]
```diff
@@ -81,6 +81,63 @@ We have a tutorial in [sherpa](https://github.com/k2-fsa/sherpa) about how
...
+        --max-duration 200 \
+```
+
+The tensorboard log is available at
+https://tensorboard.dev/experiment/QI3PVzrGRrebxpbWUPwmkA/
+
+The decoding command is:
+
+```bash
+for epoch in 72; do
```
Comment (luomingshuang) [inline egs/aishell/ASR/RESULTS.md:120]: Here, maybe you can simplify it to 
```
for m in greedy_search modified_beam_search fast_beam_search ; do
      ./pruned_transducer_stateless2/decode.py \
        --epoch 72 \
        --avg 14 \
        --exp-dir ./pruned_transducer_stateless2/exp \
        --lang-dir data/lang_char \
        --max-duration 200 \
        --decoding-method $m
done
```

Comment (TeaPoly) [reply egs/aishell/ASR/RESULTS.md]: Fixed.

File: egs/aishell/ASR/RESULTS.md [hunk 2]
```diff
@@ -81,6 +81,58 @@ We have a tutorial in [sherpa](https://github.com/k2-fsa/sherpa) about how
...
+    --lang-dir data/lang_char \
+    --max-duration 200 \
+    --decoding-method $m
+
+done
+```
+
+Pretrained models, training logs, decoding logs, and decoding results
+are available at
+<https://huggingface.co/teapoly/icefall-aishell-pruned-transducer-stateless2-2022-08-18>
```
Comment (csukuangfj) [inline egs/aishell/ASR/RESULTS.md:133]: Could you also upload a torchscript model?

You can pass `--jit 1` to `export.py` to obtain a `cpu_jit.pt` file.

**Comment by csukuangfj** [comment]
Thanks!

File: egs/aishell/ASR/RESULTS.md (hunk 2 above)
Comment (TeaPoly) [reply egs/aishell/ASR/RESULTS.md:133]: Uploaded.
